pkg/oauthserver/auth.go

```diff
@@ -356,7 +356,12 @@ func (c *OAuthServerConfig) getAuthenticationHandler(mux oauthserver.Mux, errorH
 			}
 			if identityProvider.UseAsChallenger {
 				// For now, all password challenges share a single basic challenger, since they'll all respond to any basic credentials
-				challengers["basic-challenge"] = passwordchallenger.NewBasicAuthChallenger("openshift")
+				// The separate "bootstrap" realm allows 'oc' to distinguish between bootstrap IDP and others
+				if config.IsBootstrap(identityProvider) {
+					challengers["basic-challenge"] = passwordchallenger.NewBasicAuthChallenger("bootstrap")
+				} else {
+					challengers["basic-challenge"] = passwordchallenger.NewBasicAuthChallenger("openshift")
+				}
```

Comment:
I'd prefer this:
```golang
realm := "openshift"
if config.IsBootstrap(identityProvider) {
    realm = "openshift:kubeadmin"
}
challengers["basic-challenge"] = passwordchallenger.NewBasicAuthChallenger(realm)
```

Reply:
thanks, updated

Reply:
actually, it should probably be
```golang
realm, challengerName := "openshift", "basic-challenge"
if config.IsBootstrap(identityProvider) {
    realm, challengerName = "openshift:kubeadmin", "kubeadmin-challenge"
}
challengers[challengerName] = passwordchallenger.NewBasicAuthChallenger(realm)
```
my bad

Reply:
updated

Reply:
curious, do we use/process that "basic-challenge"/"kubeadmin-challenge" anywhere? I don't see it.

Reply:
the challengers are passed into a union here: https://github.com/openshift/oauth-server/blob/a7f975a5f3dc07f270dc64275033a66308d578a2/pkg/oauthserver/auth.go#L430

What I want is to send different WWW-Authenticate headers for kubeadmin and for normal OpenShift login, but in that case we cannot overwrite the `basic-challenge` challenger with a different one

Reply:
i see now, thanks